certora/specs/GovernorStates.spec

```diff
@@ -0,0 +1,176 @@
+    assert (stateBefore != stateAfter) => (
+        stateBefore == UNSET() => (
+            stateAfter == PENDING() && f.selector == propose(address[],uint256[],bytes[],string).selector
+        ) &&
+        stateBefore == PENDING() => (
+            (stateAfter == CANCELED() && f.selector == cancel(address[],uint256[],bytes[],bytes32).selector)
+        ) &&
+        stateBefore == SUCCEEDED() => (
+            (stateAfter == QUEUED()   && f.selector == queue(address[],uint256[],bytes[],bytes32).selector) ||
+            (stateAfter == EXECUTED() && f.selector == execute(address[],uint256[],bytes[],bytes32).selector)
+        ) &&
+        stateBefore == QUEUED() => (
+            (stateAfter == EXECUTED() && f.selector == execute(address[],uint256[],bytes[],bytes32).selector)
+        ) &&
+        stateBefore == ACTIVE()    => false &&
+        stateBefore == CANCELED()  => false &&
+        stateBefore == DEFEATED()  => false &&
+        stateBefore == EXECUTED()  => false
+    );
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: State transitions caused by time passing                                                                      │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule stateTransitionWait(uint256 pId, env e1, env e2) {
+    require clockSanity(e1);
+    require clockSanity(e2);
+    require clock(e2) > clock(e1);
+
+    uint8 stateBefore = state(e1, pId);
+    uint8 stateAfter  = state(e2, pId);
+
+    assert (stateBefore != stateAfter) => (
+        stateBefore == PENDING()   => stateAfter == ACTIVE() &&
+        stateBefore == ACTIVE()    => (stateAfter == SUCCEEDED() || stateAfter == DEFEATED()) &&
+        stateBefore == UNSET()     => false &&
+        stateBefore == SUCCEEDED() => false &&
+        stateBefore == QUEUED()    => false &&
+        stateBefore == CANCELED()  => false &&
+        stateBefore == DEFEATED()  => false &&
+        stateBefore == EXECUTED()  => false
+    );
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: State corresponds to the vote timing and results                                                              │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule stateIsConsistentWithVotes(uint256 pId, env e) {
+    require clockSanity(e);
+    requireInvariant proposalStateConsistency(pId);
+
+    uint48  currentClock  = clock(e);
+    uint8   currentState  = state(e, pId);
+    uint256 snapshot      = proposalSnapshot(pId);
+    uint256 deadline      = proposalDeadline(pId);
+    bool    quorumSuccess = quorumReached(pId);
+    bool    voteSuccess   = voteSucceeded(pId);
+
+    // Pending: before vote starts
+    assert currentState == PENDING() => (
```

Comment:
We can't say much about `quorumSuccess` because the quorum might be `= 0` (in which case no vote = success) or `> 0` (in which case no vote = no success)

For the `voteSuccess` its more complex. I order to say we don't have success, we need to "prove" that the votes counters are empty. This is done by `votesImplyCreated` with is part of `GovernorInvariants.spec`. If we want to prove it here, we need to import the invariant and require it. Note that this will cause all the runs of this spec file to re-verify the invariant.